### PR TITLE
CSS-10115 Add check for policies before deletion

### PR DESCRIPTION
**Pull request by AmberCharitos**
To avoid loss of policies check for custom policies before deletion on relation broken event. Additionally:
- Improves unittesting

### CODE REVIEW

**Review by gtato — APPROVED** [review]
looks good

File: src/relations/provider.py [hunk 1]
```diff
@@ -196,14 +206,46 @@ def _create_ranger_client(self):
...
         """
         ranger = self._create_ranger_client()
         retrieved_service = ranger.get_service_by_id(service_id)
 
-        if retrieved_service is not None:
-            ranger.delete_service_by_id(service_id)
+        if retrieved_service is None:
+            return
+
+        if self._has_custom_policies(
```
Comment (mertalpt) [inline src/relations/provider.py:222]: question (blocking): Why do we care about preserving custom policies?

File: src/relations/provider.py [hunk 2]
```diff
@@ -130,14 +142,17 @@ def _create_ranger_service(self, ranger, data, event):
             ranger: ranger client
             data: relation data
             event: relation event
+
+        Returns:
+            created_service: the object for the created service
         """
         service_name = data["name"]
         retrieved_service = ranger.get_service(service_name)
         if retrieved_service is not None:
```
[the remark this answers is not in the document]
Comment (AmberCharitos) [reply src/relations/provider.py]: Updated.

**Review by mertalpt — APPROVED** [review]

File: src/relations/provider.py (hunk 1 above)
Comment (mertalpt) [reply src/relations/provider.py:230]: What happens if you cannot access the UI and you need to nuke the relation? Does `--force` work?

Comment (AmberCharitos) [reply src/relations/provider.py:230]: This change still allows you to remove the relation, and trino will no longer use ranger policies but default to it's own. The trino service will just be left in tact in Ranger for you to reconnect to if you add the relation without re-creating all the policies. Worst case scenario, if you can't access the UI, you just have a unused ranger services archive in Ranger which is providing no functionality.

**Pull request merged**